commit 7d72f7dd8d28d18c59a08a7d43bd4e0654043103
Author: Thorsten Zachmann <t.zachmann@zagge.de>
Date:   Sat Aug 4 06:42:14 2012 +0200

    Make sure not to write behind the allocated memory
    
    Validate the input data to not write behind the allocated memory. This fixes a
    buffer overflow found by Charlie Miller.
    (cherry picked from commit 8652ab672eaaa145dfb3782f5011de58aa4cc046)

diff --git a/filters/words/msword-odf/wv2/src/styles.cpp b/filters/words/msword-odf/wv2/src/styles.cpp
index c3132f0..0d691f5 100644
--- a/filters/words/msword-odf/wv2/src/styles.cpp
+++ b/filters/words/msword-odf/wv2/src/styles.cpp
@@ -248,6 +248,11 @@ throw(InvalidFormatException)
 #ifdef WV2_DEBUG_STYLESHEET
         wvlog << "cbUPX: " << cbUPX << endl;
 #endif
+        // do not overflow the allocated buffer grupx
+        if (offset + cbUPX > grupxLen) {
+            wvlog << "====> Error: grupx would overflow!" << endl;
+            return false;
+        }
         for ( U16 j = 0; j < cbUPX; ++j ) {
             grupx[ offset + j ] = stream->readU8();  // read the whole UPX
 #ifdef WV2_DEBUG_STYLESHEET
